lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,281 @@
+import           Control.Lens (makeLenses, to)
+import qualified Data.ByteString as BS
+import           Data.Default (Default (..))
+import qualified Data.Text.Buildable
+import           Formatting (Format, bprint, build, later, (%))
+import           System.Directory (doesFileExist)
+import           System.Directory (renameFile)
+import           System.FileLock (FileLock, SharedExclusive (..), lockFile, unlockFile,
+                                  withFileLock)
+import           System.FilePath (takeDirectory, takeFileName)
+import           System.IO (hClose, openBinaryTempFile)
+#ifdef POSIX
+import           System.Wlog (WithLogger)
+#endif
+import           Test.QuickCheck (Arbitrary (..))
+import           Test.QuickCheck.Arbitrary.Generic (genericArbitrary, genericShrink)
+import           Universum
+
+import           Pos.Binary.Class (Bi (..), Cons (..), Field (..), decodeFull', deriveSimpleBi,
+                                   encodeListLen, enforceSize, serialize')
+import           Pos.Crypto (PublicKey, PublicKey)
+import           Pos.Util.UserKeyError (UserPublicError (..))
+
+import           Test.Pos.Crypto.Arbitrary ()
+
+#ifdef POSIX
+import           Formatting (oct, sformat)
+import qualified System.Posix.Files as PSX
+import qualified System.Posix.Types as PSX (FileMode)
+import           System.Wlog (logWarning)
+#endif
+
+-- Because of the Formatting import
+{-# ANN module ("HLint: ignore Use fewer imports" :: Text) #-}
+
+-- | Describes HD wallets keyfile content
+data WalletUserPublic = WalletUserPublic
+    { _wupWalletName :: Text                -- ^ name of wallet
+    , _wupAccounts   :: [(Word32, Text)]    -- ^ accounts coordinates and names
+    , _wupAddrs      :: [(Word32, Word32)]  -- ^ addresses coordinates
+    } deriving (Show, Generic)
+
+instance Arbitrary WalletUserPublic where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+makeLenses ''WalletUserPublic
+
+instance Buildable WalletUserPublic where
+    build WalletUserPublic{..} =
+        bprint ("{ wallet name = "%build%
+                ", accounts = "%pairsF%", addresses = "%pairsF%" }")
+        _wupWalletName
+        _wupAccounts
+        _wupAddrs
+      where
+        pairsF :: (Buildable a, Buildable b) => Format r ([(a, b)] -> r)
+        pairsF = later $ mconcat . map (uncurry $ bprint ("("%build%", "%build%")"))
+
+deriveSimpleBi ''WalletUserPublic [
+    Cons 'WalletUserPublic [
+        Field [| _wupWalletName :: Text               |],
+        Field [| _wupAccounts   :: [(Word32, Text)]   |],
+        Field [| _wupAddrs      :: [(Word32, Word32)] |]
+    ]]
+
+-- | User public data.
+data UserPublic = UserPublic
+    { _upKeys    :: [PublicKey]
+    , _upWallet  :: Maybe WalletUserPublic
+    , _upPath    :: FilePath
+    , _upLock    :: Maybe FileLock
+    } deriving (Generic)
+
+instance Arbitrary (Maybe FileLock) => Arbitrary UserPublic where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
```

Comment:
`Arbitrary` instances do not belong on the library itself. Please see the [code organisation guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/code-organisation-guidelines.md?expand=1).